File: iOS/QGVAPlayerDemo/QGVAPlayerDemo/ViewController.m
// ViewController.m
// Tencent is pleased to support the open source community by making vap available.
//
// Copyright (C) 2020 THL A29 Limited, a Tencent company.  All rights reserved.
//
// Licensed under the MIT License (the "License"); you may not use this file except in
// compliance with the License. You may obtain a copy of the License at
//
// http://opensource.org/licenses/MIT
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
// either express or implied. See the License for the specific language governing permissions and
// limitations under the License.

#import <Masonry/Masonry.h>
#import "ViewController.h"
#import "UIView+VAP.h"
#import "QGVAPWrapView.h"
#import "QGBaseAnimatedImageFrame.h"
#import "QGMP4AnimatedImageFrame.h"
#import <AVFoundation/AVFoundation.h>

@interface ViewController () <HWDMP4PlayDelegate, VAPWrapViewDelegate>

@property (nonatomic, strong) UIButton *vapButton;
@property (nonatomic, strong) UIButton *vapxButton;
@property (nonatomic, strong) UIButton *vapWrapViewButton;

@property (nonatomic, strong) VAPView *vapView;

@end

@implementation ViewController

//日志接口
void qg_VAP_Logger_handler(VAPLogLevel level, const char* file, int line, const char* func, NSString *module, NSString *format, ...) {
    
    if (format.UTF8String == nil) {
        NSLog(@"log包含非utf-8字符");
        return;
    }
    if (level > VAPLogLevelDebug) {
        va_list argList;
        va_start(argList, format);
        NSString* message = [[NSString alloc] initWithFormat:format arguments:argList];
        file = [NSString stringWithUTF8String:file].lastPathComponent.UTF8String;
        NSLog(@"<%@> %s(%@):%s [%@] - %@",@(level), file, @(line), func, module, message);
        va_end(argList);
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupAudioSession];
    
    //日志
    [UIView registerHWDLog:qg_VAP_Logger_handler];
    
    //vap-经典效果
    _vapButton = [[UIButton alloc] initWithFrame:CGRectMake(0, 100, CGRectGetWidth(self.view.frame), 90)];
    _vapButton.backgroundColor = [UIColor lightGrayColor];
    [_vapButton setTitle:@"电竞方案（退后台结束）" forState:UIControlStateNormal];
    [_vapButton addTarget:self action:@selector(playVap) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:_vapButton];
    
    //vapx-融合效果
    _vapxButton = [[UIButton alloc] initWithFrame:CGRectMake(0, CGRectGetMaxY(_vapButton.frame)+60, CGRectGetWidth(self.view.frame), 90)];
    _vapxButton.backgroundColor = [UIColor lightGrayColor];
    [_vapxButton setTitle:@"融合特效（退后台暂停/恢复）" forState:UIControlStateNormal];
    [_vapxButton addTarget:self action:@selector(playVapx) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:_vapxButton];
    
    //使用WrapView，支持ContentMode
    _vapWrapViewButton = [[UIButton alloc] initWithFrame:CGRectMake(0, CGRectGetMaxY(_vapxButton.frame)+60, CGRectGetWidth(self.view.frame), 90)];
    _vapWrapViewButton.backgroundColor = [UIColor lightGrayColor];
    [_vapWrapViewButton setTitle:@"WrapView-ContentMode" forState:UIControlStateNormal];
    [_vapWrapViewButton addTarget:self action:@selector(playVapWithWrapView) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:_vapWrapViewButton];
}

- (void)setupAudioSession {
    AVAudioSession* avsession = [AVAudioSession sharedInstance];
    NSError *error = nil;
    if (![avsession setCategory:AVAudioSessionCategoryPlayback withOptions:0 error:&error]) {
        if (error) NSLog(@"AVAudioSession setCategory failed : %ld, %s", (long)error.code, [error.localizedDescription UTF8String]);
        return;
    }
    if (![avsession setActive:YES error:&error]) {
        if (error) NSLog(@"AVAudioSession setActive failed : %ld, %s", (long)error.code, [error.localizedDescription UTF8String]);
    }
}

#pragma mark - 各种类型的播放

- (void)playVap {
    VAPView *mp4View = [[VAPView alloc] initWithFrame:CGRectMake(0, 0, 752/2, 752/2)];
    //默认使用metal渲染，使用OpenGL请打开下面这个开关
//    mp4View.hwd_renderByOpenGL = YES;
    mp4View.center = self.view.center;
    
    [self.view addSubview:mp4View];
    [mp4View mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.top.right.bottom.mas_equalTo(0);
    }];
    mp4View.userInteractionEnabled = YES;
    mp4View.hwd_enterBackgroundOP = HWDMP4EBOperationTypeStop;
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(onImageviewTap:)];
    [mp4View addGestureRecognizer:tap];
    NSString *resPath = [NSString stringWithFormat:@"%@/Resource/demo.mp4", [[NSBundle mainBundle] resourcePath]];
    //单纯播放的接口
    //[mp4View playHWDMp4:resPath];
    //指定素材混合模式，重复播放次数，delegate的接口
    
    //注意若素材不含vapc box，则必须用调用如下接口设置enable才可播放
    [mp4View enableOldVersion:YES];
    [mp4View playHWDMP4:resPath repeatCount:-1 delegate:self];
}

//vap动画
- (void)playVapx {
    NSString *mp4Path = [NSString stringWithFormat:@"%@/Resource/vap.mp4", [[NSBundle mainBundle] resourcePath]];
    VAPView *mp4View = [[VAPView alloc] initWithFrame:self.view.bounds];
    [self.view addSubview:mp4View];
    [mp4View mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.top.right.bottom.mas_equalTo(0);
    }];
    mp4View.center = self.view.center;
    mp4View.userInteractionEnabled = YES;
    mp4View.hwd_enterBackgroundOP = HWDMP4EBOperationTypePauseAndResume; // ⚠️ 建议设置该选项时对机型进行判断，屏蔽低端机
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(onImageviewTap:)];
    [mp4View addGestureRecognizer:tap];
    [mp4View setMute:YES];
    [mp4View playHWDMP4:mp4Path repeatCount:-1 delegate:self];
    
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [mp4View pauseHWDMP4];
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [mp4View resumeHWDMP4];
        });
    });
}

/// 使用WrapView，支持ContentMode
- (void)playVapWithWrapView {
    static BOOL pause = NO;
    QGVAPWrapView *wrapView = [[QGVAPWrapView alloc] initWithFrame:self.view.bounds];
    wrapView.center = self.view.center;
    wrapView.contentMode = QGVAPWrapViewContentModeAspectFill;
    wrapView.autoDestoryAfterFinish = NO;
    [self.view addSubview:wrapView];
    [wrapView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.top.right.bottom.mas_equalTo(0);
    }];
    NSString *resPath = [NSString stringWithFormat:@"%@/Resource/vap.mp4", [[NSBundle mainBundle] resourcePath]];
    [wrapView setMute:NO];
    [wrapView playHWDMP4:resPath repeatCount:0 delegate:self];
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(doNothingonImageviewTap:)];
    
    __weak __typeof(wrapView) weakWrapView = wrapView;
    [wrapView addVapGesture:tap callback:^(UIGestureRecognizer *gestureRecognizer, BOOL insideSource, QGVAPSourceDisplayItem *source) {
        if ((pause = !pause)) {
            [weakWrapView pauseHWDMP4];
        } else {
            [weakWrapView resumeHWDMP4];
        }
    }];
}

#pragma mark -  mp4 hwd delegate

#pragma mark -- 播放流程
- (void)viewDidStartPlayMP4:(VAPView *)container {
    
}

- (void)viewDidFinishPlayMP4:(NSInteger)totalFrameCount view:(UIView *)container {
    //note:在子线程被调用
    NSLog(@"totalFrameCount -- %ld",totalFrameCount);
}

- (void)viewDidPlayMP4AtFrame:(QGMP4AnimatedImageFrame *)frame view:(UIView *)container {
    //note:在子线程被调用
    NSLog(@"frame -- %@",frame);
}

- (void)viewDidStopPlayMP4:(NSInteger)lastFrameIndex view:(UIView *)container {
    //note:在子线程被调用
    NSLog(@"lastFrameIndex -- %ld",lastFrameIndex);
    dispatch_async(dispatch_get_main_queue(), ^{
        [container removeFromSuperview];
    });
}

- (BOOL)shouldStartPlayMP4:(VAPView *)container config:(QGVAPConfigModel *)config {
    return YES;
}

- (void)viewDidFailPlayMP4:(NSError *)error {
    NSLog(@"%@", error.userInfo);
}




#pragma mark -- 融合特效的接口 vapx

//provide the content for tags, maybe text or url string ...
- (NSString *)contentForVapTag:(NSString *)tag resource:(QGVAPSourceInfo *)info {
    
    NSDictionary *extraInfo = @{@"[sImg1]" : @"http://shp.qlogo.cn/pghead/Q3auHgzwzM6GuU0Y6q6sKHzq3MjY1aGibIzR4xrJc1VY/60",
                                @"[textAnchor]" : @"我是主播名",
                                @"[textUser]" : @"我是用户名😂😂",};
    return extraInfo[tag];
}

//provide image for url from tag content
- (void)loadVapImageWithURL:(NSString *)urlStr context:(NSDictionary *)context completion:(VAPImageCompletionBlock)completionBlock {
    
    //call completionBlock as you get the image, both sync or asyn are ok.
    //usually we'd like to make a net request
    dispatch_async(dispatch_get_main_queue(), ^{
        UIImage *image = [UIImage imageNamed:[NSString stringWithFormat:@"%@/Resource/qq.png", [[NSBundle mainBundle] resourcePath]]];
        //let's say we've got result here
        completionBlock(image, nil, urlStr);
    });
}

#pragma mark - gesture

- (void)onImageviewTap:(UIGestureRecognizer *)ges {
    
    [ges.view removeFromSuperview];
}

- (void)doNothingonImageviewTap:(UIGestureRecognizer *)ges {
    
}

#pragma mark - WrapViewDelegate

//provide the content for tags, maybe text or url string ...
- (NSString *)vapWrapview_contentForVapTag:(NSString *)tag resource:(QGVAPSourceInfo *)info {
    NSDictionary *extraInfo = @{@"[sImg1]" : @"http://shp.qlogo.cn/pghead/Q3auHgzwzM6GuU0Y6q6sKHzq3MjY1aGibIzR4xrJc1VY/60",
                                @"[textAnchor]" : @"我是主播名",
                                @"[textUser]" : @"我是用户名😂😂",};
    return extraInfo[tag];
}

//provide image for url from tag content
- (void)vapWrapView_loadVapImageWithURL:(NSString *)urlStr context:(NSDictionary *)context completion:(VAPImageCompletionBlock)completionBlock {
    dispatch_async(dispatch_get_main_queue(), ^{
        UIImage *image = [UIImage imageNamed:[NSString stringWithFormat:@"%@/Resource/qq.png", [[NSBundle mainBundle] resourcePath]]];
        completionBlock(image, nil, urlStr);
    });
}


- (void)vapWrap_viewDidStartPlayMP4:(VAPView *)container {
    
}
- (void)vapWrap_viewDidPlayMP4AtFrame:(QGMP4AnimatedImageFrame*)frame view:(VAPView *)container {
    QGMP4AnimatedImageFrame * frameModel = frame;
    
    NSLog(@"frame.frameIndex -- %ld",frameModel.frameIndex);
    NSLog(@"frame.duration -- %d",frameModel.duration);
    if (frameModel.frameIndex >= 150) {
        [container pauseHWDMP4];
    }
    
}
- (void)vapWrap_viewDidStopPlayMP4:(NSInteger)lastFrameIndex view:(VAPView *)container {
    //note:在子线程被调用
    NSLog(@"lastFrameIndex -- %ld",lastFrameIndex);
    dispatch_async(dispatch_get_main_queue(), ^{
        [container removeFromSuperview];
    });
}
- (void)vapWrap_viewDidFinishPlayMP4:(NSInteger)totalFrameCount view:(VAPView *)container {
    NSLog(@"totalFrameCount -- %ld",totalFrameCount);
    
}
@end

File: iOS/QGVAPlayerDemoSwift/QGVAPlayerDemoSwift/ViewController.swift
// ViewController.swift
// Tencent is pleased to support the open source community by making vap available.
//
// Copyright (C) 2020 Tencent.  All rights reserved.
//
// Licensed under the MIT License (the "License"); you may not use this file except in
// compliance with the License. You may obtain a copy of the License at
//
// http://opensource.org/licenses/MIT
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
// either express or implied. See the License for the specific language governing permissions and
// limitations under the License.

import UIKit
import QGVAPlayer

class ViewController: UIViewController, VAPWrapViewDelegate {
    var vapView = QGVAPWrapView.init()
    let vapButton = UIButton()
    var customFont: UIFont?
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        vapButton.frame = CGRect(x: 0, y: 100, width: self.view.frame.width, height: 90)
        vapButton.backgroundColor = UIColor.lightGray
        vapButton.setTitle("融合动效（更多示例请查看OC版本）", for: UIControl.State.normal)
        vapButton.addTarget(self, action: #selector(playVapx), for: UIControl.Event.touchUpInside)
        self.view.addSubview(vapButton)
        self.view.addSubview(vapView)
        QGVAPTextureLoader.loadCustomFont(nil)
        QGVAPTextureLoader.loadCustomFont { fontSize, isBold in
            return UIFont.systemFont(ofSize: 12)
        }
    }
    
    @objc func playVapx() {
        
        let mp4Path = String.init(format: "%@/Resource/vap.mp4", Bundle.main.resourcePath!)
        vapView.center = self.view.center
        vapView.isUserInteractionEnabled = true
        vapView.contentMode = .top
        vapView.hwd_enterBackgroundOP = HWDMP4EBOperationType.stop
        //        let gesture = UITapGestureRecognizer.init(target: self, action: #selector(onTap(gesture:)))
        //        vapView.addGestureRecognizer(gesture)
        vapView.playHWDMP4(mp4Path, repeatCount: 0, delegate: self)
        
        
        //        test.ttf
        //    // 使用示例：从文档目录加载
            if let docPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {
                let fontPath = docPath.appendingPathComponent("test.ttf").path
                if let font = loadFontFromFile(path: fontPath, size: 16) {
                    self.customFont = font
                }
            }
    }
    
    func loadFontFromFile(path: String, size: CGFloat) -> UIFont? {
        guard let fontURL = URL(fileURLWithPath: path) as CFURL?,
              let dataProvider = CGDataProvider(url: fontURL),
              let cgFont = CGFont(dataProvider) else {
            return nil
        }
        let fontName = cgFont.postScriptName! as String
        if self.isFontRegistered(fontName: fontName) {
            return UIFont(name: fontName, size: size)
        }
        
        var error: Unmanaged<CFError>?
        // 注册字体到系统
        if CTFontManagerRegisterGraphicsFont(cgFont, &error) {
            
            return UIFont(name: fontName, size: size)
        } else {
            print("字体注册失败：\(error?.takeUnretainedValue() ?? NSError() as! CFError)")
            return nil
        }
    }
    
    /// 检查字体是否已注册
    func isFontRegistered(fontName: String) -> Bool {
        // 遍历所有字体家族
        for family in UIFont.familyNames {
            // 获取家族下的所有字体名称
            for name in UIFont.fontNames(forFamilyName: family) {
                if name == fontName {
                    return true
                }
            }
        }
        return false
    }
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        vapView.frame = self.view.bounds
    }
    @objc func onTap(gesture: UIGestureRecognizer) {
        gesture.view?.stopHWDMP4()
        gesture.view?.removeFromSuperview()
    }
    

    
    func vapWrapview_content(forVapTag tag: String, resource info: QGVAPSourceInfo) -> String {
        let extraInfo: [String:String] = ["[sImg1]" : "http://shp.qlogo.cn/pghead/Q3auHgzwzM6GuU0Y6q6sKHzq3MjY1aGibIzR4xrJc1VY/60",
                                          "[textAnchor]" : "afdas123123123",
                                          "[textUser]" : "afdaf12312312",]
        
        return extraInfo[tag] ?? ""
    }

    func vapWrapView_loadVapImage(withURL urlStr: String, context: [AnyHashable : Any], completion completionBlock: @escaping VAPImageCompletionBlock) {
        DispatchQueue.main.async {
            let image = UIImage.init(named: String.init(format:"%@/Resource/qq.png", Bundle.main.resourcePath!))
            completionBlock(image, nil, urlStr)
        }
    }
    
    func vapWrapView_loadVapContent(_ content: String, context: [AnyHashable : Any]) -> UIImage? {
        let resoure = context["resource"] as? QGVAPSourceInfo;
        let color = resoure?.color ?? UIColor.white
        let size = resoure?.size ?? CGSize.zero
        let isBlod = resoure?.style == QGAGAttachmentSourceStyle.boldText;
        QGVAPTextureLoader.drawingCustomImage(forText: content, color: color, size: size, bold: isBlod) { fontSize, isBlod in
             
            if let docPath = Bundle.main.path(forResource: "test.ttf", ofType: nil) {
                if let font = self.loadFontFromFile(path: docPath, size: fontSize) {
                    return font
                }
            }
            if isBlod {
                return UIFont.boldSystemFont(ofSize: fontSize)
            }
            return UIFont.systemFont(ofSize: fontSize)
        }
        return nil
    }
    
}

